fix(transformer): `wrap_in_arrow_function_iife` take span of input `Expression`

overlookmotel opened the pull request:
#8530 introduced a small change. Previously the IIFE had the span of the original expression which is being wrapped, but after that PR it got a dummy `SPAN`. Fix that.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8547?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8547** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8547?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8547?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8546** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8546?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8545** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8545?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8530** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8530?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8529** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8529?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8511** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8511?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8547).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
I don't know how to get the source map to work, I have a question about this, do we need to reset the span of input `Expression`?

overlookmotel commented:
> I don't know how to get the source map to work, I have a question about this, do we need to reset the span of input `Expression`?

I have been wondering exactly the same thing. I don't know the answer. I would *hope* that it's OK not to reset the inner span, and that when the source map is generated it should see it doesn't need to create a separate entry for the inner span because it's redundant information.

My logic is that when printing a large file which has not been transformed at all, the source map should say "Line 1: Same as before. Line 2: Same as before. ..." and not create a massive source map saying "Line 1 column 0 -> Line 1 column 0. Line 1 column 1 -> Line 1 column 1. Line 1 column 2 -> Line 1 column 2. ... and so on and so on".

So, similarly, you'd hope that it can handle an `Expression` containing another `Expression` where both have the same `Span`, and figure out that the inner span can be ignored.

If it doesn't, we should probably look at fixing that in the source map generator!

overlookmotel commented:
PS: I'm treating spans as "best effort" at the moment. I'm just working on the simple principle that if you create a node which replaces another node, the new node should have the span of the old one.

But I don't *really* understand how source maps work, and I think we need some proper testing infrastructure to figure it out properly.

Dunqing commented:
> PS: I'm treating spans as "best effort" at the moment. I'm just working on the simple principle that if you create a node which replaces another node, the new node should have the span of the old one.
> 

I am also following the same principle now. 

> But I don't _really_ understand how source maps work, and I think we need some testing infrastructure to figure it out properly.

Yes, we are still thinking it should work like this, we need evidence to tell us we are doing it right. Not far from figuring out how it works.

overlookmotel committed:
fix(transformer): `wrap_in_arrow_function_iife` take span of input `Expression` (#8547)

#8530 introduced a small change. Previously the IIFE had the span of the original expression which is being wrapped, but after that PR it got a dummy `SPAN`. Fix that.
